src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3205,15 +3205,18 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           //         the corresponding synthetics to the package class, only to the package object class.
           // scala/bug#6734 Locality test below is meaningless if we're not even in the correct tree.
           //         For modules that are synthetic case companions, check that case class is defined here.
+          // scala/bug#10783 ditto for synthetic companions of derived value classes.
           def shouldAdd(sym: Symbol): Boolean = {
+            def classDefinedHere(s: Symbol): Boolean = stats exists {
+              case t: ClassDef => t.symbol eq s
```

Comment:
Not sure if it matters, but this will miss `ClassDefs`-s that are wrapped in `DocDef`-s when running under scalacheck mode. The painful ease with which one can make this mistake is lamented in https://github.com/scala/scala/blob/9acab45aeeadef2f63da69faf81465cc15599789/src/compiler/scala/tools/nsc/ast/TreeInfo.scala#L76-L107

Reply:
A scaladoc test should already demonstrate the existing issue.